plutus-contract/src/Plutus/Contract/Test/ContractModel.hs

```diff
@@ -0,0 +1,1015 @@
+
+-- | Get the contract-specific part of the model state.
+--
+--   `Spec` monad update functions: `$=` and `$~`.
+contractState :: Getter (ModelState state) state
+contractState = contractStateL
+
+-- | Get the current slot.
+--
+--   `Spec` monad update functions: `wait` and `waitUntil`.
+currentSlot :: Getter (ModelState state) Slot
+currentSlot = currentSlotL
+
+-- | Get the current wallet balances. These are delta balances, so they start out at zero and can be
+--   negative. The absolute balances used by the emulator can be set in the `CheckOptions` argument
+--   to `propRunActionsWithOptions`.
+--
+--   `Spec` monad update functions: `withdraw`, `deposit`, `transfer`.
+balances :: Getter (ModelState state) (Map Wallet Value)
+balances = balancesL
+
+-- | Get the current balance for a wallet. This is the delta balance, so it starts out at zero and
+--   can be negative. The absolute balance used by the emulator can be set in the `CheckOptions`
+--   argument to `propRunActionsWithOptions`.
+--
+--   `Spec` monad update functions: `withdraw`, `deposit`, `transfer`.
+balance :: Wallet -> Getter (ModelState state) Value
+balance w = balancesL . at w . non mempty
+
+-- | Get the amount of tokens forged so far. This is used to compute `lockedValue`.
+--
+--   `Spec` monad update functions: `forge` and `burn`.
+forged :: Getter (ModelState state) Value
+forged = forgedL
+
+-- | How much value is currently locked by contracts. This computed by subtracting the wallet
+--   `balances` from the `forged` value.
+lockedValue :: ModelState s -> Value
+lockedValue s = s ^. forged <> inv (fold $ s ^. balances)
+
+-- | Monads with read access to the model state: the `Spec` monad used in `nextState`, and the `DL`
+--   monad used to construct test scenarios.
+class Monad m => GetModelState m where
+    -- | The contract state type of the monad. For both `Spec` and `DL` this is simply the @state@
+    --   parameter of the respective monad.
+    type StateType m :: *
+
+    -- | Get the current model state.
+    getModelState :: m (ModelState (StateType m))
+
+-- | Get the contract state part of the model state.
+getContractState :: GetModelState m => m (StateType m)
+getContractState = _contractState <$> getModelState
+
+-- | Get a component of the model state.
+askModelState :: GetModelState m => (ModelState (StateType m) -> a) -> m a
+askModelState f = f <$> getModelState
+
+-- | Get a component of the contract state.
+askContractState :: GetModelState m => (StateType m -> a) -> m a
+askContractState f = askModelState (f . _contractState)
+
+-- | Get a component of the model state using a lens.
+viewModelState :: GetModelState m => Getting a (ModelState (StateType m)) a -> m a
+viewModelState l = askModelState (^. l)
+
+-- | Get a component of the contract state using a lens.
+viewContractState :: GetModelState m => Getting a (StateType m) a -> m a
+viewContractState l = viewModelState (contractStateL . l)
```

Comment:
The general theme of my feeling reading this section was: surely we could get away with more standard lens stuff rather than having to define our own variants of lots of them? We're already exposing the user to lenses a bit , so that seems maybe preferable. Perhaps I'd have to try it and see if it was bad!

Reply:
The problem I see is that there isn't a standard type class for a monad that support read access to some value. Both `MonadState` and `MonadReader` require that you also let the user change the value.